### PR TITLE
[Driver][SYCL][FPGA] Do not optimize after wrapping an object

### PR DESCRIPTION
**Pull request by mdtoguchi**
When performing -fsycl-link=early and generating an archive for
FPGA, the host object ended up being truncated in the archive.
After wrapping the host object (which generates a .bc file), that
wrapped object goes through an additional compilation step. When
enabling optimizations (-O2, etc), the wrapped host object would
be optimized away, causing problems when going from the early
device to either exe or image.  Do not pass the optimization option
during this specific step.

### CODE REVIEW

File: clang/test/Driver/sycl-offload-intelfpga.cpp [hunk 1]
```diff
@@ -38,7 +38,8 @@
 // CHK-FPGA-LINK: clang-offload-wrapper{{.*}} "-o=[[WRAPOUT:.+\.bc]]" "-host=x86_64-unknown-linux-gnu" {{.*}} "-kind=sycl"
 // CHK-FPGA-LINK: llc{{.*}} "-o" "[[OBJOUTDEV:.+\.o]]" "[[WRAPOUT]]"
 // CHK-FPGA-EARLY: clang-offload-wrapper{{.*}} "-host" "x86_64-unknown-linux-gnu" "-o" "[[WRAPOUTHOST:.+\.bc]]" "-kind=host"
-// CHK-FPGA-EARLY: clang{{.*}} "-o" "[[OBJOUT:.+\.o]]" {{.*}} "[[WRAPOUTHOST]]"
+// CHK-FPGA-EARLY-NOT: clang{{.*}} "-O2"
```
Comment (AGindinson) [inline clang/test/Driver/sycl-offload-intelfpga.cpp:41]: Minor: extending the check with the target could be useful

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -5151,13 +5162,17 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   // preprocessed inputs and configure concludes that -fPIC is not supported.
   Args.ClaimAllArgs(options::OPT_D);
 
+  bool SkipO =
+      Args.hasArg(options::OPT_fsycl_link_EQ) && ContainsWrapperAction(&JA);
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/Clang.cpp:5166]: The fact that early optimizations are disabled for FPGA is the reason why the change isn't going to result in device-side optimization disabling when compiling from source. Does this statement stand?

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:5166]: This will not impact device side optimization due to the fact that wrapping never occurs before the device compilation.  The search for wrapper recursively checks inputs and does not check other adjacent toolchains.  This should only impact items after wrapping which in this case is only the host object wrapping for FPGA.

**Review by AGindinson — APPROVED** [review]
LGTM! The remaining comment is a minor one, I'll re-approve if it gets addressed.

**Review by AGindinson — APPROVED** [review]

File: clang/test/Driver/sycl-offload-intelfpga.cpp (hunk 1 above)
Comment (mdtoguchi) [reply clang/test/Driver/sycl-offload-intelfpga.cpp]: Done :)
